speechbrain/lobes/models/huggingface_transformers/whisper.py

```diff
@@ -308,65 +280,329 @@ def _pad_or_trim(self, array, axis=-1):
+                index=torch.arange(length, device=array.device),
             )
 
-        if array.shape[axis] < self._n_samples:
+        if array.shape[axis] < length:
             pad_widths = [(0, 0)] * array.ndim
             pad_widths[axis] = (
                 0,
-                self._n_samples - array.shape[axis],
+                length - array.shape[axis],
             )
             array = nn.functional.pad(
                 array, [pad for sizes in pad_widths[::-1] for pad in sizes]
             )
 
         return array
 
-    def forward_decoder(self, audio_features, decoder_input_ids):
+    def forward_encoder(self, mel):
+        """Takes an input mel and return its corresponding encoder states.
+        Returns the last hidden state of the encoder or all hidden states if
+        output_all_hiddens is True.
+
+        Arguments
+        ---------
+        mel : torch.Tensor (signal)
+            A batch of audio mel to transform to features.
+
+        Returns
+        -------
+        torch.Tensor
+            The last hidden state of the encoder or all hidden states if
+            output_all_hiddens is True.
+        """
+        encoder_states = self.model.encoder(
+            mel, output_hidden_states=self.output_all_hiddens
+        )
+        if self.output_all_hiddens:
+            return torch.stack(encoder_states.hidden_states)
+        else:
+            return encoder_states.last_hidden_state
+
+    def forward_decoder(
+        self,
+        encoder_states,
+        decoder_input_ids,
+        use_cache=True,
+        past_key_values=None,
+    ):
         """Perform one step of the whisper decoder.
 
         Arguments
         ---------
-        audio_features : torch.Tensor
-            A batch of audio features (mel + whisper encoding).
+        encoder_states : torch.Tensor
+            A batch of encoder_states features (mel + whisper feature extractor).
         decoder_input_ids : torch.Tensor
-            A batch of decoder inputs tokens.
-            The first tokens need to dictate the behavior of the decoder.
-            It needs to start with the bos_token, the language token,
-            the task token, and finally the timestamp token.
-
+            Input tokens for the decoder. This can be language, task, etc.
             Please refer to the whisper paper for more details or go to the
             seq2seq2.py file in SpeechBrain to see how to generate the tokens
             with Greedy Search and/or Beam Search.
+        use_cache : bool
+            If True, keys and values are returned as output for KV caching.
+        past_key_values : torch.Tensor (default: None)
+            If not None, the past key values are used for KV caching and
+            avoid recomputing the attention weights.
 
         Returns
         -------
         logits : torch.Tensor
-            Outputs of decoder
+            The logits of the decoder.
         attn : torch.Tensor
```

Comment:
Should be marked as ` | None` in the docstring, and it should be mentioned that disabling `output_attentions` results in this being `None`.